Feat/explain predictions

slebastard opened the pull request:
Added a function `explain(sf_dataset: SFDataset, conf: ModuleType)` that explains the relative contribution of each input feature to the logistic regression declared in `conf.MODEL_PIPELINE`. The function takes as input a SFDataset containing the dataset predictions were ran on, and outputs a SFDataset with two additional columns:
- a column "micro_expl", containing a dictionary with the relative contribution of each individual feature to the failure risk score
- a columns "macro_expl", containing a dictionary with the relative contribution of each group of features, declared in `conf.FEATURE_GROUPS`, to the failure risk score

Note: in the future, adding a test of explain() in our e2e testing may be interesting:
- testing that all elements of conf.FEATURES are reprensented in dictionary "micro_expl"
- testing that all feature groups of conf.FEATURE_GROUPS are represented in "macro_expl"
- testing that the explanation weights are between -1 and 1

slebastard committed:
feat: adding function explain, called on predict set

slebastard committed:
feat: added feature groups objects for existing models

slebastard committed:
feat: intermediary version of explainer

slebastard committed:
Adding aggregated explanations to explain method

slebastard committed:
Turned explain columns into dict, merged with base df

slebastard committed:
Updating explain()'s docstring

slebastard committed:
Removing backslashs from explain()'s docstring

vviers commented on models/default/model_conf.py:
C'est quand même assez redondant avec notre définition des FEATURES
Je pense qu'on peut faire mieux ! par exemple parser les valeurs de `FEATURES` automatiquement depuis `FEATURE_GROUPS` :

```python
features = []
for group, sublist_of_features in FEATURE_GROUP.items():
    features.expand(sublist_of_feature)
```

ça évite de devoir maintenir 2 listes de features

fcoufour replied:
Ca donne néanmoins la possibilité d'avoir des features n'appartenant à aucun groupe

slebastard replied:
Suivant notre discussion, je vous propose de garder uniquement `FEATURE_GROUPS` dans son format actuel de dictionnaire indexé par des noms de groupe, de le renommer `FEATURES`, et de créer un groupe `miscellaneous` par les features qui ne feront partie d'aucun groupe. Il n'y en a pour l'instant aucun.

slebastard replied:
Modification décrite ci-dessus effectuée

vviers commented on predictsignauxfaibles/__main__.py:
```diff
@@ -117,6 +122,109 @@ def evaluate(
     return {"balanced_accuracy": balanced_accuracy, "fbeta": fbeta}
 
 
+def explain(
```
Move this crucial function to its own module `predictsignauxfaibles.explainability` then import it here

slebastard replied:
Good idea! Will do

slebastard replied:
Done, ready for review

slebastard committed:
feat: Defining FEATURES from FEATURE_GROUPS

slebastard committed:
feat: moved explain() fct to new explicability module

slebastard committed:
feat: adding a miscellaneous feature group

slebastard committed:
fix: removing useless imports

slebastard committed:
fix: adding final newline

vviers commented on predictsignauxfaibles/explicability.py:
```diff
@@ -0,0 +1,108 @@
+    ]
+    flat_data = pd.DataFrame(sf_data.data[[feat for (group, feat) in multi_columns]])
+    data = pd.DataFrame(sf_data.data[[feat for (group, feat) in multi_columns]])
+    data.columns = multi_columns
```
Je pense que tu n'as pas besoin de cette ligne ? tu réassignes une nouvelle valeur à `data.columns` juste en dessous

slebastard replied:
Soit, mais le traitement `data.columns` dans la seconde ligne utilise l'état de `data.columns` tel qu'il est mis à jour à la première ligne.
C'est du code bien fonctionnel mais je pense que je pourrai faire plus compact. J'envisage de faire une refacto à terme.

slebastard replied:
Effectivement, j'ai modifié ça pour utiliser directement `multi_columns` dans le 2nd statement

vviers commented on predictsignauxfaibles/explicability.py:
```diff
@@ -0,0 +1,108 @@
+                        feat,
+                    ]
+                ]
+            )
```
c'est hyper bizarre ce que le formatteur à fait là 😄

vviers reviewed (dismissed):
LGTM ! Merci pour ce super taf ! 💯 

Peut-être en bonus commenter un poil plus les étapes de ton code pour faciliter la reprise par qqun d'autre, mais en ouvrant python sur le côté pour tester des bouts de code ça se comprend bien 🤓

slebastard committed:
feat: adding feature for explanibility radar

slebastard committed:
feat: adding feature to compute list of variables to interface in front-end

slebastard committed:
fix: FEATURE's definition from FEATURE_GROUP

slebastard committed:
feat: adding new explainability columns to exported csv

slebastard committed:
feat: adding logging of explainability (takes a few seconds)

slebastard committed:
fix: removing offset from radar vars & reversing score def

slebastard commented:
May 11th, 2021 updates:
- Moved function `explain()` to new `predictsignauxfaibles.explainability`module
- In `predictsignauxfaibles.explainability`, added column `macro_radar` containing scores between 0 (very bad situation, high risk) and 1 (very good situation, low risk) to interface a radar plot in the front-end
- In `predictsignauxfaibles.explainability`, added columns `expl_selection` containing two lists: `select_concerning` (respectively `select_reassuring`) containing the list of variables that have a substantially positive/bad (resp. negative/good) contribution to the risk score. This selection was based on a nice statistical criterion on the maximal hypothetical contribution of a group of features to the global score
- In `models/default/model_conf.py` and `models/small/model_conf.py`, fixed a bug on the definition of `FEATURES` from `FEATURE_GROUPS`
@vviers could you please take one last look at the modifications above?

slebastard commented:
Here is a theoretical ressource to understand how radar scores were computed:
![sf_logreg_explradar_theory](https://user-images.githubusercontent.com/15195227/117858064-c4fd4e80-b28d-11eb-9854-05a65b89b9f4.png)

slebastard committed:
fix: merging redundant operations on data.columns

vviers reviewed (dismissed):
Top, merci ! 

(par contre tu as un conflit dans utils.py qu'il faut résoudre avant de pouvoir merger)

slebastard committed:
fix: Removing model offset from selected features for full-text explanations

slebastard commented:
Merci pour ta review! Je m'occupe du conflit sur `utils.py`

slebastard committed:
feat: Moving  from  to

slebastard committed:
Merge branch 'develop' into feat/explain_predictions
